src/Microsoft.ApplicationInsights/TelemetryClient.cs

```diff
@@ -516,44 +518,95 @@ public void Initialize(ITelemetry telemetry)
             }
 
             telemetry.Context.Initialize(this.Context, instrumentationKey);
-            foreach (ITelemetryInitializer initializer in this.configuration.TelemetryInitializers)
+
+            ISupportAdvancedSampling telemetryWithSampling = telemetry as ISupportAdvancedSampling;
+
+            // Telemetry can be already sampled out if that decision was made before calling Track()
+            bool sampledOut = telemetryWithSampling?.IsProactivelySampledOut ?? false;
+            bool shouldTryHeadSampling = this.proactiveSamplingEnabled && (telemetryWithSampling?.SupportsProactiveSampling ?? false);
+
+            if (!sampledOut)
             {
-                try
+                if (shouldTryHeadSampling)
                 {
-                    initializer.Initialize(telemetry);
+                    // If telemetry was not sampled out, we need to check its sampling score after we learn its Operation Id                    
+                    bool shouldCheckSamplingScore = true;
+
+                    for (int index = 0; index < this.configuration.TelemetryInitializers.Count; index++)
+                    {
+                        if (shouldCheckSamplingScore && !string.IsNullOrEmpty(telemetry.Context.Operation.Id))
+                        {
+                            if (SamplingScoreGenerator.GetSamplingScore(telemetry.Context.Operation.Id) >= this.configuration.GetLastObservedSamplingPercentage(telemetryWithSampling.ItemTypeFlag))
+                            {
+                                telemetryWithSampling.IsProactivelySampledOut = true;
+                                sampledOut = true;
+                                break;
+                            }
+                            else
+                            {
+                                shouldCheckSamplingScore = false;
```

Comment:
is there an implicit assumption that sampling will be based on the operation.Id set by the first telemetryinitializer, and that operation.id set by any later initializers will not be taken into consideration for sampling calculations?

Reply:
Ok - if not already done, can you make it obvious in the PR description so there is a trail of decisions made.. (or maybe a comment in the code itself)